python-manual/modules/ROOT/pages/transactions.adoc

```diff
@@ -0,0 +1,306 @@
+
+.A transaction with multiple queries, client logic, and potential roll backs.
+[source,python]
+----
+from neo4j import GraphDatabase
+
+
+URI = "neo4j://localhost"
+AUTH = ("neo4j", "secret")
+employee_threshold=10
+
+
+def main():
+    with GraphDatabase.driver(URI, auth=AUTH) as driver:
+        with driver.session(database="neo4j") as session:
+            for i in range(100):
+                name = "Thor"+str(i)
+                org_id = session.execute_write(employ_person_tx, name)
+                print(f"User {name} added to organization {org_id}")
+
+
+def employ_person_tx(tx, name):
+    # Create new Person node with given name, if not existing already
+    result = tx.run("""
+        MERGE (p:Person {name: $name})
+        RETURN p.name AS name
+        """, name=name
+    )
+
+    # Obtain most recent organization ID and the number of people linked to it
+    result = tx.run("""
+        MATCH (o:Organization)
+        RETURN o.id AS id, COUNT{(p:Person)-[r:WORKS_FOR]->(o)} AS employees_n
+        ORDER BY o.created_date DESC
+        LIMIT 1
+    """)
+    org = result.single()
+
+    if org is not None and org["employees_n"] == 0:
+        raise Exception("Most recent organization is empty.")
+        # Transaction will roll back -> not even Person is created!
+
+    # If org does not have too many employees, add this Person to that
+    if org is not None and org.get("employees_n") < employee_threshold:
+        result = tx.run("""
+            MATCH (o:Organization {id: $org_id})
+            MATCH (p:Person {name: $name})
+            MERGE (p)-[r:WORKS_FOR]->(o)
+            RETURN $org_id AS id
+            """, org_id=org["id"], name=name
+        )
+
+    # Otherwise, create a new Organization and link Person to it
+    else:
+        from neo4j.time import Date
+        result = tx.run("""
+            MATCH (p:Person {name: $name})
+            CREATE (o:Organization {id: randomuuid(), created_date: datetime()})
+            MERGE (p)-[r:WORKS_FOR]->(o)
+            RETURN o.id AS id
+            """, name=name, date=Date.today()
+        )
+
+    # Return the Organization ID to which the new Person ends up in
+    return result.single()["id"]
+
+
+if __name__ == "__main__":
+    main()
+----
+
+Should a transaction fail for a reason that the driver deems transient, it will automatically retry to run the transaction function (with an exponentially increasing delay). For this reason, *transaction functions must be _idempotent_* (i.e., they should produce the same effect when run several times), because you do not know upfront how many times they are going to be executed. In practice, this means that you should not edit nor rely on globals, for example. Note that although transactions functions might be executed multiple times, the queries inside it will always run only once.
```

Comment:
```suggestion
Should a transaction fail for a reason that the driver deems transient, it automatically retries to run the transaction function (with an exponentially increasing delay). 
For this reason, *transaction functions must be _idempotent_* (i.e., they should produce the same effect when run several times), because you do not know upfront how many times they are going to be executed. 
In practice, this means that you should not edit nor rely on globals, for example. 
Note that although transactions functions might be executed multiple times, the queries inside it are always run only once.
```